core/src/test/scala/integration/kafka/api/TransactionsTest.scala

```diff
@@ -0,0 +1,314 @@
+          producer.send(producerRecord(topic2, key, value, willBeCommitted = shouldCommit))
+        }
+
+        producer.sendOffsetsToTransaction(offsetsToCommit(consumer), consumerGroupId)
+        if (shouldCommit) {
+          producer.commitTransaction()
+          recordsProcessed += records.size
+          debug(s"committed transaction.. Last committed record: ${new String(records.last.value(), "UTF-8")}. Num " +
+            s"records written to $topic2: $recordsProcessed")
+        } else {
+          producer.abortTransaction()
+          trace("aborted transaction")
+          // reset the subscription. Practically, we will have a new consumer instance when transactions are aborted
+          // (like during a streams re balance). But we need to do this manually for this test, otherwise the consumer
+          // will continue to fetch from the cached fetch position and the test will fail because it will fail to
+          // produce the expected number of messages to the destination topic since some of the aborted transactions
+          // will not result in offset resets.
+          resetToLastCommittedPosition(consumer)
+        }
+      }
+    } catch {
+      case e : Exception =>
+        fail ("Received an unexpected exception during the 'consume-process-produce' loop", e)
+    } finally {
+      producer.close()
+      consumer.close()
+    }
+
+    // Inspite of random aborts, we should still have exactly 1000 messages in topic2. Ie. we should not
+    // re-copy or miss any messages from topic1, since the consumed offsets were committed transactionally.
+    val verifyingConsumer = transactionalConsumer("foobargroup")
+    verifyingConsumer.subscribe(List(topic2))
+    val valueSeq = pollUntilAtLeastNumRecords(verifyingConsumer, 1000).map { record =>
+      assertCommittedAndGetValue(record).toInt
+    }
+    verifyingConsumer.close()
+    val valueSet = valueSeq.toSet
+    assertEquals(s"Expected 1000 values in $topic2.", 1000, valueSeq.size)
+    assertEquals(s"Expected ${valueSeq.size} unique messages in $topic2.", valueSeq.size, valueSet.size)
+  }
+
+  @Test
+  def testFencing() = {
+    val transactionalId = "my-t.id"
+    val producer1 = transactionalProducer(transactionalId)
+    val producer2 = transactionalProducer(transactionalId)
+    val consumer = transactionalConsumer()
+    consumer.subscribe(List(topic1, topic2))
+
+    try {
+      producer1.initTransactions()
+
+      producer1.beginTransaction()
+      producer1.send(producerRecord(topic1, "1", "1", willBeCommitted = false))
+      producer1.send(producerRecord(topic2, "3", "3", willBeCommitted = false))
+
+      producer2.initTransactions()  // ok, will abort the open transaction.
+      producer2.beginTransaction()
+      producer2.send(producerRecord(topic1, "2", "4", willBeCommitted = true))
+      producer2.send(producerRecord(topic2, "2", "4", willBeCommitted = true))
+
+      try {
+        producer1.commitTransaction()
```

Comment:
I added test cases to check for fencing on send and also test fencing on add partitions.